# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: docshell/shistory/nsSHEntry.h
# Commit: 6ebc7618a5fa
# Full Hash: 6ebc7618a5fadb59585a4d8b9f17eb65e5654cc6
# Author: Anny Gakhokidze <agakhokidze@mozilla.com>
# Date: 2020-03-10 21:44:54
# Regressor Bug: 1546759
# File Overlap Count: 2
# Description:
#   Bug 1546759 - Change nsSHistory::WalkHistoryEntries to walk browsing context tree instead of doc shell tree, r=peterv,nika
#   
#   WalkHistoryEntries function gets called by nsSHistory::CloneAndReplaceChild
#   and nsSHistory::SetChildHistoryEntry recursively, so those have to be moved
#   into the parent process. This eliminates many sync IPC calls.
# ==============================================================================

diff -r ed75364b23c3 -r 6ebc7618a5fa docshell/shistory/nsSHEntry.h
--- a/docshell/shistory/nsSHEntry.h	Tue Mar 10 11:39:06 2020 +0000
+++ b/docshell/shistory/nsSHEntry.h	Tue Mar 10 14:28:22 2020 +0000
@@ -28,6 +28,17 @@
 class nsIURI;
 class nsIReferrerInfo;
 
+// Structure for passing around entries via XPCOM from methods such as
+// nsSHistory::CloneAndReplaceChild, nsSHistory::SetChildHistoryEntry and
+// nsSHEntry::SyncTreesForSubframeNavigation, that need to swap entries in
+// docshell, to corresponding Recv methods so that we don't have to create
+// actors until we are about to return from the parent process
+struct EntriesAndBrowsingContextData {
+  nsCOMPtr<nsISHEntry> oldEntry;
+  nsCOMPtr<nsISHEntry> newEntry;
+  mozilla::dom::BrowsingContext* context;
+};
+
 class nsSHEntry : public nsISHEntry {
  public:
   NS_DECL_ISUPPORTS
@@ -37,6 +48,11 @@
 
   static nsresult Startup();
   static void Shutdown();
+  void SyncTreesForSubframeNavigation(
+      uint64_t aOtherPid, nsISHEntry* aEntry,
+      mozilla::dom::BrowsingContext* aTopBC,
+      mozilla::dom::BrowsingContext* aIgnoreBC,
+      nsTArray<EntriesAndBrowsingContextData>* aEntriesToUpdate);
 
  protected:
   explicit nsSHEntry(mozilla::dom::SHEntrySharedParentState* aState);